### ISSUE TITLE
Quarkus CLI doesn't honor platform version for codestarts

### ISSUE THREAD

**Issue by rsvoboda** [quarkusio/quarkus#25384]
### Describe the bug

Quarkus CLI doesn't honor platform version for codestarts

running `quarkus create app app --platform-bom=io.quarkus.platform:quarkus-bom:2.7.5.Final -x=quarkus-resteasy-jackson,quarkus-spring-web` pulls in codestart relevant to the quarkus cli version, not the codestart relevant to the platform version.
Generated project ends up with both `io.quarkus:quarkus-resteasy-reactive-jackson` and `io.quarkus:quarkus-resteasy-jackson` and doesn't compile in the end.

spring-web-codestart defines io.quarkus:quarkus-resteasy-jackson in 2.7.5.Final so `io.quarkus:quarkus-resteasy-reactive-jackson` shouldn't be pulled in. 

https://github.com/quarkusio/quarkus/blob/2.7.5.Final/independent-projects/tools/base-codestarts/src/main/resources/codestarts/quarkus/extension-codestarts/spring-web-codestart/codestart.yml
 - io.quarkus:quarkus-resteasy-jackson

https://github.com/quarkusio/quarkus/blob/main/independent-projects/tools/base-codestarts/src/main/resources/codestarts/quarkus/extension-codestarts/spring-web-codestart/codestart.yml
 - io.quarkus:quarkus-resteasy-reactive-jackson

Generating project using maven command works and codestart based on 2.7.5.Final is used / `io.quarkus:quarkus-resteasy-reactive-jackson` doesn't get pulled in
```
mvn io.quarkus.platform:quarkus-maven-plugin:2.7.5.Final:create \
    -DprojectGroupId=org.acme \
    -DprojectArtifactId=getting-started \
    -DclassName="org.acme.getting.started.GreetingResource" \
    -DplatformVersion=2.7.5.Final -Dextensions=quarkus-resteasy-jackson,quarkus-spring-web
```

### Expected behavior

codestart based on 2.7.5.Final is used

### Actual behavior

codestart based on CLI version is used

### How to Reproduce?

CLI way fails to run:
```
quarkus create app app --platform-bom=io.quarkus.platform:quarkus-bom:2.7.5.Final -x=quarkus-resteasy-jackson,quarkus-spring-web
mvn clean verify -f app/
```

```
grep jackson app/pom.xml
      <artifactId>quarkus-resteasy-jackson</artifactId>
      <artifactId>quarkus-resteasy-reactive-jackson</artifactId>
```
```
quarkus --version
2.8.2.Final
```

Maven way works:
```
mvn io.quarkus.platform:quarkus-maven-plugin:2.7.5.Final:create \
    -DprojectGroupId=org.acme \
    -DprojectArtifactId=getting-started \
    -DclassName="org.acme.getting.started.GreetingResource" \
    -DplatformVersion=2.7.5.Final -Dextensions=quarkus-resteasy-jackson,quarkus-spring-web
mvn clean verify -f getting-started/
```

### Output of `uname -a` or `ver`

macOS Monterey

### Output of `java -version`

Java 17

### Quarkus version or git rev

2.8.2.Final

**Comment by ia3andy** [comment]
cc @aloubyansky

**Comment by ia3andy** [comment]
@maxandersen maybe that's the reason why our cross-version tests are almost all passing, we need to add some checks on the version in the generated project.

**Comment by ebullient** [comment]
IIRC (without looking, and I haven't looked recently), the code starts were pulling whatever was latest, or whatever was built into the CLI (for the codestart itself). Parameter passing though the various abstraction layers is a lot more complicated than it should be, so I wouldn't have been surprised if it was dropped (and we didn't have a way to test well against other versions to verify it wasn't..). Good thing we have an environment for testing this now.. :)

**Comment by ia3andy** [comment]
> IIRC (without looking, and I haven't looked recently), the code starts were pulling whatever was latest, or whatever was built into the CLI (for the codestart itself). Parameter passing though the various abstraction layers is a lot more complicated than it should be, so I wouldn't have been surprised if it was dropped (and we didn't have a way to test well against other versions to verify it wasn't..). Good thing we have an environment for testing this now.. :)

As made by @aloubyansky the codestarts loading is directed by the platform descriptor which contains the versioned artifacts.

This codestart (spring-web) though is part of the base codestarts which means we use the ones shipped with the generator, so what I wonder is how is mvn working.

**Comment by ia3andy** [comment]
Per design, changing a base codestarts affect all generated apps independent of the selected platform, they shouldn't break past versions..

Base codestarts are more coupled to the generator so we keep them together. The system would let us load them by platform, but then we should introduce some compatibility checks and possibly sometime break compat (https://github.com/quarkusio/quarkus/issues/25321).

**Comment by ia3andy** [comment]
I've started a zulip thread to discuss the problem of adding extension from codestarts which always lead to problems: https://quarkusio.zulipchat.com/#narrow/stream/187038-dev/topic/Adding.20extensions.20from.20codestarts/near/281293196

**Comment by rsvoboda** [comment]
> using a previous version of maven which ships with an old version of the codestarts

I guess you mean "previous version of quarkus-maven-plugin"

With mvn command I can say "give me everything 2.7.5.Final based"

But with Quarkus CLI I can only get the BOM/extensions 2.7.5.Final based, but not the codestarts.
And with the pace of changes that Quarkus has this is a bit problematic. There should be a way to get the codestarts related to the platform version.

Workaround is to use 2.7.5 based quarkus-cli, but who wants to install/maintain different versions of the quarkus cli.
And for RHBQ 2.7 this will be quite a big trouble because Quarkus 2.8 moved to RESTEasy Reactive as the default and codestarts changed.

**Comment by ia3andy** [comment]
@rsvoboda this is a bug, you should be able to use the latest CLI to create working app for any previous platform.

The issue comes from this: https://github.com/quarkusio/quarkus/issues/23265#issuecomment-1096438592 which we fixed with a side-effect bug (this one)

**Comment by ia3andy** [comment]
This was fixed in #26698
